### PR TITLE
WC-1308 Reduce force-deletes unless explicitly confirmed

### PR DESCRIPTION
**Pull request by matthewdavidrodgers**
Changes the default ?force param to the delete API call to be false.

Also uses the /scripts/{scriptName}/references and /tails/by-consumer/{scriptName} endpoints to determine what resources depend on the current script. We check for other scripts that use this one as outbound worker (for dynamic dispatch), as a tail worker, as a service binding, or use a durable object implemented by this script. If any other resources rely on this worker, we require another step of confirmation, describing the dependencies, and only then do we allow doing a force delete.

for WC-1308

**Author has included the following, where applicable:**

- [x] Tests
- [x] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

### CODE REVIEW

**Comment by matthewdavidrodgers** [comment]
Definitely open to any suggestions on copy here - see the included test for an example

File: .changeset/twelve-pumpkins-crash.md
Comment (penalosa) [inline .changeset/twelve-pumpkins-crash.md:5]: Could this have a bit more detail?

File: packages/wrangler/src/delete.ts
Comment (penalosa) [inline packages/wrangler/src/delete.ts:217]: Could the "script" language be replaced with "worker"?

File: packages/wrangler/src/delete.ts
Comment (penalosa) [inline packages/wrangler/src/delete.ts:133]: Can this reuse types from https://github.com/cloudflare/workers-sdk/blob/a9349a89296dc98ac8fc52dbb013f298c9596d8f/packages/wrangler/src/worker.ts?

File: packages/wrangler/src/delete.ts
Comment (penalosa) [inline packages/wrangler/src/delete.ts:202]: This is a fairly complex conditional—could it be refactored into multiple variables/lifted out into a separate function?

File: packages/wrangler/src/delete.ts
Comment (penalosa) [inline packages/wrangler/src/delete.ts:164]: Is this intermediate string status needed? What about `true | false | undefined`, corresponding to the `force` parameter value?

File: packages/wrangler/src/delete.ts
Comment (penalosa) [inline packages/wrangler/src/delete.ts:94]: What kind of error response does the API return in the `"needed_denied"` case?

File: packages/wrangler/src/delete.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/delete.ts]: refacored out into helper fns

File: packages/wrangler/src/delete.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/delete.ts]: moved types to the shared `worker.ts`

File: packages/wrangler/src/delete.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/delete.ts]: updated language from "script" to "Worker"

File: packages/wrangler/src/delete.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/delete.ts]: Depends on what you mean - in the "needed_denied" case, we don't make any call to the API. The user has rejected the confirmation for using extra aggressive force-deletes, so we return early.
But if you're asking what would happen if we tried to delete with `?force=false` when the Worker is a dependent service, then the API would return a Forbidden error with a message describing which resource is blocking the delete.

**Review by penalosa — APPROVED** [review]

**Comment by penalosa** [comment]
Actually, looking at this again—the delete command should take a `--force` flag to override the checking behaviour

**Comment by lrapoport-cf** [comment]
> Actually, looking at this again—the delete command should take a `--force` flag to override the checking behaviour

@penalosa should merge be blocked on this? if so, can you update your approval :)

**Review by penalosa — CHANGES_REQUESTED** [review]
This should have a `--force` flag

**Review by penalosa — APPROVED** [review]
Looks good to me. Let's have another person in @cloudflare/wrangler take a pass before merging though

File: packages/wrangler/src/__tests__/delete.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/delete.test.ts:363]: If the user specifies `--force` I would expect that we should not even ask this question...

File: packages/wrangler/src/__tests__/delete.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/delete.test.ts:445]: Are these new endpoints live and stable in EWC?

File: packages/wrangler/src/deployment-bundle/worker.ts
Comment (petebacondarwin) [inline packages/wrangler/src/deployment-bundle/worker.ts:289]: Given that they are only used in the `delete.ts` file, there is no need to have these types here.
In fact, going forward, the aim is that this file describes what is uploaded as part of a deployment, rather than responses from API requests.
Can we move them into `delete.ts`, please?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Nice PR! Really clean and understandable code.

Can you confirm that the new requests are to stable public APIs? So that Wrangler won't get broken by them being changed or removed.

I think it would be good to only ask about deleting if there is no `--force` flag, and also can we move the new types into the `delete.ts` file.

Otherwise LGTM!

File: packages/wrangler/src/__tests__/delete.test.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/__tests__/delete.test.ts:438]: they are indeed - these have been out and stable since pre-dev week, references for much longer

**Review by petebacondarwin — APPROVED** [review]
❤️

**Comment by petebacondarwin** [comment]
Um last thought... is this a breaking change for users that previously expected their deletions to be forced through, even if that is bad for them?

**Comment by matthewdavidrodgers** [comment]
> Um last thought... is this a breaking change for users that previously expected their deletions to be forced through, even if that is bad for them?

erm, I don't know. I suppose you could classify it as a breaking change, albeit not a very impactful one

**Comment by petebacondarwin** [comment]
OK so I think the opinion is that this does not deserve a major versio bump.
Looks good to me. Let's :shipit:

**Pull request merged**
